### ISSUE TITLE
Cooling v Luminosity revamp, including bug in bound-free photon generation

### ISSUE THREAD

**Issue by jhmatthews** [sirocco-rt/sirocco#261]
We have discovered a major issue with bound-free photon generation, in that we currently use lum_fb as our "bound-free wind luminosity" -- i.e., actual radiative energy emitted by the wind, but lum_fb is calculated from total_fb which gives a "bound-free cooling rate" -- i.e. energy transferred from the thermal pool. 

@kslong , myself and @Higginbottom are going to work to improve the code both in terms of clarity and commenting, but also we are generally going to distinguish explicitly between cooling rates and radiative luminosities. As a result, new variables are required in the geo and plasma structures, and we will probably n

The actual bug in bf photons will require that lum_fb gets the integrated bound-free emissivity from

A few examples:
* lum_adiabatic should really be cool_adiabatic
* lum_fb should give the radiative luminosity, cool_fb should give the cooling rate
* lum_ff and cool_ff would be the same. It is not clear yet if we should have both variables in this case.

This work is being carried out in the low_temp branch. Nick is going to start working on the process of converting variables. I will be responsible for making sure everything works with macro-atoms. We should try and document here.  

This is pretty much priority A for me Knox and Nick at the moment as it needs to be fixed before new meaningful science can be done.

**Comment by kslong** [comment]
@Higginbottom has now updated the structure names, though since he has done this via sed  a lot of comment updates are needed.  Knox has created a new set of routines for cooling in a new file called cooling.c and updated the total_emission to call fb_total in FB_FULL mode.  The revised version of the program on the low_temp branch runs but produces identical results to what was produced before any of the structure updates.

Knox believes the problem is likely associated with the fact that we store results in a set of fb arrays and if the frequency range of interest is the same as the last time fb_total was called, then it uses the values stored in the arrays instead of calculating them with a new mode.  However, this needs to be verified.  More generally, we need to think how these arrays are created and used.  The original justification for storing the arrays was that calculating them was thought to be time consuming, but I suspect that may not be a good justification any long.  In any event, some thought should go into this by @jhmatthews and the rest of us.

**Comment by Higginbottom** [comment]
Storing the arrays is still definitely needed. When I implemented my inner shell rate calculations (to compute dielectronic recombination) the code ran exceptionally slowly before I included a new array for inner shell recombinations. I believe that the code works are required, and I can't see an obviously better way of doing this.
I have therefore renamed the two arrays fbstruc.cool and fbstruc.inner_cool and created a new array fbstruc.lum which is populated by integrals using the fb_choice option FB_FULL. 
This has fixed the problem above, and I believe we are now getting wind luminosities as required. With James's StarCoolTests.pf file, I get an order of magnitude more rr luminosity. It makes little difference to the results however because (I think) the wind is not optically thick enough, so the star luminosity is still much much brighter than the wind and so dominates.
I will take a look at KSLs optically thick runs to see if anything is changing. 
Other than that, I've pulled in the changes, so low_temp should now run with the new code.

**Comment by Higginbottom** [comment]
Excitingly, running Knox's stellar.pf now converges. Oddly, the converged temperature is 70000K, whilst the star has a surface temperature of 40000K. 

Here are the various heating/cooling/luminosity data from the end of the run. We can see that the wind luminosity is now dominated by recombination... But is it right? Can the wind end up *hotter* than the star?
Oh, and it takes 5x longer to run...

!!wind_update: Absorbed flux    1.16e+37  (photo 5.54e+36 ff 5.57e+36 compton 2.48e+33 auger 8.96e+26 induced_compton 9.51e+31 lines 5.23e+35)
!!wind_update: Wind heating     1.16e+37  (photo 5.54e+36 ff 5.57e+36 compton 2.48e+33 auger 8.96e+26 induced_compton 9.51e+31 lines 5.23e+35 adiabatic 0.00e+00)
!!wind_update: Wind luminosity  3.55e+37 (recomb 3.11e+37 ff 4.45e+36 lines 3.41e+34) after update
!!wind_update: Wind cooling     1.16e+37 (recomb 5.59e+36 ff 4.45e+36 compton 1.92e+33 DR 1.49e+36 DI 7.21e+34 lines 3.41e+34 adiabatic 0.00e+00) after update

**Comment by Higginbottom** [comment]
Interesting to note the number of photons made from each source.
By the last cycle, less than 1% of the photons in each band come from the star. So maybe we end up in a situation where the radiation field in the cell is *totally* dominated by its own radiation. Maybe this means that if it ends up hot for whatever reason, it will be very hard for it to cool down....

CK and I were discussing this the other day - and if we want an optically thick cell to reach LTE, you can essentially produce and reabsorb a vast number of wind photons, totally dominating the number of photons you want to make.

**Comment by jhmatthews** [comment]
Nick, is the 70000K radiation or electron temperature? and what is the corresponding other temperature?

**Comment by kslong** [comment]
This is good news, Nick.  Is the observed luminosity about equal to input luminosity from the star?  That is a crucial test.

**Comment by Higginbottom** [comment]
70 000K is the electron temperature, the radiation temperature in the cell is logged as 94 000K

Radiated luminosity is reported as 8.37e36, the total flux from the star, adding together all the bands is 1.82e35.

The total luminosity before transphot, so all the photons made by the star and the wind is equal to 5.57e37!

**Comment by jhmatthews** [comment]
But what we need is the total **escaping** luminosity, right? because that luminosity is counted multiple times? Say I have N cells that are each optically thick, they could reprocess all the flux from the star and I would have to produce a wind luminosity of N x Lstar

**Comment by kslong** [comment]
Yes. That is what I meant by the observed luminosity, and I think that is the number Nick was quoting.

**Comment by Higginbottom** [comment]
I have rerun Knox's model using the PL model for ionization. This gives a converged temperature of 31 000K, and it converges *fast*.
I guess that the spectrum in the cell is not well modelled by a black body, even though it is being illuminated by one.

radiated luminosity is 2e36, with the same star luminosity, 1.8e35

**Comment by Higginbottom** [comment]
There seems to be a problem with the wind save file. Can't open in in py_wind. Get an error that suggests something is not being dimensioned correctly. Grrrrr

**Comment by Higginbottom** [comment]
Following my #266 pull request, I can now see the absorbed photon flux.

For the run above, 

Absorbed flux = 1.3e37
Wind Heating = 7.87e36
wind cooling = 7.87e36 (so converged)
Wind Luminosity 1.7e37

so, there is 4e36 worth of wind cooling that is not reabsorbed by the wind.
Presumbably 50% of this goes outwards and 50% goes inwards

This tallies with the radiated luminosity - 2e36 (plus a piffling 2e35 from the star).

So, our problem here is that we are generating *lots* of wind luminosity, and *hoping* it will all be reabsorbed cancelling itself out and leaving the feeble glow from the star. 

Or, we are making too much rr luminosity....

**Comment by kslong** [comment]
Might want to work with a model With smaller r and use boundary layer to avoid hitting the star

Sent from my iPhone
Th
On Jun 23, 2017, at 4:44 AM, Higginbottom <notifications@github.com<mailto:notifications@github.com>> wrote:


Following my #266<https://github.com/agnwinds/python/pull/266> pull request, I can now see the absorbed photon flux.

For the run above,

Absorbed flux = 1.3e27
Wind Heating = 7.87e36
wind cooling = 7.87e36 (so converged)
Wind Luminosity 1.7e37

so, there is 4e36 worth of wind cooling that is not reabsorbed by the wind.
Presumbably 50% of this goes outwards and 50% goes inwards

This tallies with the radiated luminosity - 2e36 (plus a piffling 2e35 from the star).

So, our problem here is that we are generating lots of wind luminosity, and hoping it will all be reabsorbed cancelling itself out and leaving the feeble glow from the star.

Or, we are making too much rr luminosity....

**Comment by Higginbottom** [comment]
I've put some figures up here:

https://sites.google.com/site/agnwinds/system/app/pages/createPage?source=/home

**Comment by Higginbottom** [comment]
Interestingly, if I *increase* the massloss rate to 1e-5, I end up with the wind at 40000K, 

Absorbed flux = 1.73e39
Wind heating = 1.12e39
Wind Cooling = 1.12e39
Wind Luminosity = 1.74e39

Radiated luminosity = 3.7e36, so still more than the star makes., but looking at those huge fluxes made and absorbed in the wind it would be surprising if the balanced perfectly. But, with the higher density we are capturing almost all (as a fraction) of the generated wind photons, and so the wind is ending up at the correct temperature. Mind you - there are a *tiny* number of photons actually coming from the star  - only about 100 out of 100 000 are made in the star....

If I go up to a massloss rate of 1e-4, then because the wind emission goes as the square of the density, you end up with no photons made in the star at all. Somehow the wind still ends up at about 34 000K..

I think this demonstrates that the idea is working. But how can we properly test things?

**Comment by kslong** [comment]
It might be worthwhile to try a identical model with 100 cells. This might produce a temperature gradient and lower observed luminosity.

K

Sent from my iPhone

On Jun 23, 2017, at 10:30 AM, Higginbottom <notifications@github.com<mailto:notifications@github.com>> wrote:


Interestingly, if I increase the massloss rate to 1e-5, I end up with the wind at 40000K,

Absorbed flux = 1.73e39
Wind heating = 1.12e39
Wind Cooling = 1.12e39
Wind Luminosity = 1.74e39

Radiated luminosity = 3.7e36, so still more than the star makes., but looking at those huge fluxes made and absorbed in the wind it would be surprising if the balanced perfectly. But, with the higher density we are capturing almost all (as a fraction) of the generated wind photons, and so the wind is ending up at the correct temperature. Mind you - there are a tiny number of photons actually coming from the star - only about 100 out of 100 000 are made in the star....

If I go up to a massloss rate of 1e-4, then because the wind emission goes as the square of the density, you end up with no photons made in the star at all. Somehow the wind still ends up at about 34 000K..

I think this demonstrates that the idea is working. But how can we properly test things?

**Comment by kslong** [comment]
@Higginbottom @jhmatthews @lazygun37   I have completed a reemplementation of pdf_gen_from_array with some changes to recomb that allows one to send pdf_gen_from array multiple levels.  

What I did is as follows:

In recomb, the routine now sorts the levels in to ascending order as pdf_gen_from_array expects and in the process eliminated duplicate jumps which arose because many of our level files have more than one entry at the same ionization level, e.g for the n=2 level of hydrogen as it appears in topbase.

In pdf_gen_from array, I made changes that assure incorporate into the input array (the pdf) all of the jumps.  I set the value of the pdf at the jump to be the value just below that in the input array.  This effectively assumes that our jumps are all positive as you increase in frequency, as is the case for recombination.  After calculating the cdf of the input array, I made sure than in regridding onto a coarser grid that (a) we never interpolate except at the end, and (b) that the point of every jump was in the coarser grid.  

I believe NIck is considering ensuring that the the pdf at every jump is incorporated into the input array as calculated in recomb.c.  This is a good idea, and should work with the routines I have written without the need to change pdf.c 

I've tested this in a couple of situations and the results seemed reasonable, but obviously a lot more testing is needed.  Note that when one runs with a normal abundance model, the number of jumps is fairly large, at least 30 in some examples I have run.  This may argue for increasing the overall size of the pdf from its current value of 200.

Aside: I have not tried to deal with the problem that the level populations should go toward LTE. As several of us discussed on Friday, this will not be particular easy to carry out.  it is not simply that one needs to solve the rate equations, but that one must actually calculate the the recombination rates to individual levels.  Hopefully, this will not be necessary for our immediate problem.  I do note that it is not obvious to me, what approximation is currently being used for this in most of our modes.  This is worth investigating.
